File: frontend/flutter/ecar_app/lib/models/vehicle.dart
class Vehicle {
  final int id;
  final int customerId;
  final String brand;
  final String model;
  final int year;
  final String licensePlate;
  final String? vin;
  final int? currentMileage;
  final double? averageDailyUsage;
  final DateTime? nextServiceDueDate;
  final int? daysUntilNextService;
  
  Vehicle({
    required this.id,
    required this.customerId,
    required this.brand,
    required this.model,
    required this.year,
    required this.licensePlate,
    this.vin,
    this.currentMileage,
    this.averageDailyUsage,
    this.nextServiceDueDate,
    this.daysUntilNextService,
  });
  
  factory Vehicle.fromJson(Map<String, dynamic> json) {
    return Vehicle(
      id: json['id'],
      customerId: json['customer_id'],
      brand: json['brand'],
      model: json['model'],
      year: json['year'],
      licensePlate: json['license_plate'],
      vin: json['vin'],
      currentMileage: json['current_mileage'],
      averageDailyUsage: json['average_daily_usage']?.toDouble(),
      nextServiceDueDate: json['next_service_due_date'] != null 
          ? DateTime.parse(json['next_service_due_date']) 
          : null,
      daysUntilNextService: json['days_until_next_service'],
    );
  }
  
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'customer_id': customerId,
      'brand': brand,
      'model': model,
      'year': year,
      'license_plate': licensePlate,
      'vin': vin,
      'current_mileage': currentMileage,
      'average_daily_usage': averageDailyUsage,
    };
  }
} 

File: backend/app/models/vehicle.rb
class Vehicle < ApplicationRecord
  belongs_to :customer
  has_many :repairs, dependent: :destroy
  
  validates :brand, presence: true
  validates :model, presence: true
  validates :year, presence: true, numericality: { only_integer: true, greater_than: 1900, less_than_or_equal_to: -> { Date.current.year + 1 } }
  validates :license_plate, presence: true, uniqueness: true
  validates :vin, uniqueness: true, allow_blank: true
  validates :current_mileage, numericality: { only_integer: true, greater_than_or_equal_to: 0 }, allow_nil: true
  validates :average_daily_usage, numericality: { greater_than_or_equal_to: 0 }, allow_nil: true
  
  def next_service_due_date
    last_repair = repairs.where.not(next_service_estimate: nil).order(completion_date: :desc).first
    last_repair&.next_service_estimate
  end
  
  def days_until_next_service
    return nil if next_service_due_date.nil?
    (next_service_due_date - Date.current).to_i
  end
end
